File: java/datastructures/lib/src/main/java/datastructures/queue/Queue.java
package datastructures.queue;

public class Queue<T>{
  Node<T> front = null;
  Node<T> rear = null;


  public void enqueue(T value) {
    Node<T> newNode = new Node<>(value);
    if (this.front == null) {
      this.front = newNode;
    } else {
      this.rear.next = newNode;
    }
    this.rear = newNode;
  }

  public T dequeue() {
    T node;
    if(this.isEmpty()){
     throw new IllegalArgumentException("Queue is empty");
   }
   node = this.front.value;
   this.front = this.front.next;
   if (this.isEmpty()){
     this.rear = null;
   } return node;
  }

  public T peek() {
    if(this.isEmpty()) {
      throw new IllegalArgumentException("Queue is empty");
    } return this.front.value;
  }

 public boolean isEmpty(){
    return this.front == null;}
}

File: javascript/code-challenges/stack-and-queue/src/queue.ts
export class Queue<T> {

  storage: T[] = [];

  get size(): number {
    return this.storage.length;
  }

  get peek(): T {
    if(this.isEmpty()){
      throw new Error("Queue is empty");
    } else {
      return this.storage[this.storage.length - 1];
    }
  }

  enqueue(value: T): void {
    this.storage.unshift(value);
  }

  dequeue(): T | undefined {
    if(this.isEmpty()) {
      throw new Error("Queue is empty");
    } else {
      return this.storage.pop();
    }
  }

  isEmpty(): boolean {
    if(this.storage.length === 0){
      return true;
    } else {
      return false;
    }
  }
}
